libsnark/gadgetlib1/gadgets/pairing/mnt/mnt_pairing_params.hpp

```diff
@@ -85,21 +87,50 @@ template<> class pairing_selector<libff::mnt6_pp>
 
     typedef Fp4_variable<FqkT> Fqk_variable_type;
     typedef Fp4_mul_gadget<FqkT> Fqk_mul_gadget_type;
-    typedef Fp4_mul_gadget<FqkT> Fqk_special_mul_gadget_type;
     typedef Fp4_sqr_gadget<FqkT> Fqk_sqr_gadget_type;
 
     typedef libff::mnt4_pp other_curve_type;
 
+    typedef G1_checker_gadget<libff::mnt6_pp> G1_checker_type;
+    typedef G2_checker_gadget<libff::mnt6_pp> G2_checker_type;
+
+    typedef mnt_G1_precomputation<libff::mnt6_pp> G1_precomputation_type;
+    typedef mnt_G2_precomputation<libff::mnt6_pp> G2_precomputation_type;
+    typedef mnt_precompute_G1_gadget<libff::mnt6_pp> precompute_G1_gadget_type;
+    typedef mnt_precompute_G2_gadget<libff::mnt6_pp> precompute_G2_gadget_type;
+
     typedef mnt_e_over_e_miller_loop_gadget<libff::mnt6_pp>
         e_over_e_miller_loop_gadget_type;
     typedef mnt_e_times_e_over_e_miller_loop_gadget<libff::mnt6_pp>
         e_times_e_over_e_miller_loop_gadget_type;
+    typedef mnt_e_times_e_times_e_over_e_miller_loop_gadget<libff::mnt6_pp>
+        e_times_e_times_e_over_e_miller_loop_gadget_type;
     typedef mnt6_final_exp_gadget<libff::mnt6_pp> final_exp_gadget_type;
+};
+
+// Parameters internal to the mnt code
+template<typename ppT> class mnt_pairing_params;
```

Comment:
If the goal of this class is to hold internal pairing params to select in the mnt pairing gadget, shall that be in an `internal` namespace?

Reply:
TBH, I'd be keen to think a bit about namespaces first.  There are several other classes (and possibly functions) as well as this which could be considered "internal", so we'd probably want to move all of those at the same time  (Note, the `internal` namespace does not seem to be used in this repo yet AFAICT).  Moreover, although I've used the word "internal" in the comments here, we might want to think about which classes are truly "internal" and which are simply "mnt-specific".

Having said that, if you feel strongly I'm happy to create the `internal` namespace and move some classes there for now.

Reply:
No worries, happy for this to be done in another PR, keeping this one focused on the code migration and giving us more time to think about this `internal` namespace.